File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/screens/checkout/CheckoutViewModel.kt
package com.sukashawarma.customer.ui.screens.checkout

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sukashawarma.customer.data.CartLine
import com.sukashawarma.customer.data.CartStore
import com.sukashawarma.customer.data.Repository
import com.sukashawarma.customer.data.api.CartItemPayload
import com.sukashawarma.customer.data.api.CartProblemDto
import com.sukashawarma.customer.data.api.GatewayError
import com.sukashawarma.customer.data.api.GatewayResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.math.roundToLong

data class CheckoutState(
    val memuat: Boolean = true,
    val galat: GatewayError? = null,
    val baris: List<CartLine> = emptyList(),
    /** Semua nilai uang di bawah berasal dari gateway, bukan hitungan sendiri. */
    val subtotal: Long? = null,
    val potongan: Long? = null,
    val total: Long? = null,
    val masalah: List<CartProblemDto> = emptyList(),
    val alasan: String? = null,
    val pesanPenolakan: String? = null,
    val keranjangKosong: Boolean = false
) {
    /** Boleh lanjut membayar hanya kalau gateway benar-benar meloloskannya. */
    val bolehLanjut: Boolean
        get() = !memuat && galat == null && total != null && masalah.isEmpty() && alasan == null
}

fun CartLine.kePayload() = CartItemPayload(
    menuItemId = menuItemId,
    name = nama,
    unitPrice = hargaSatuan.toDouble(),
    quantity = jumlah,
    note = catatan
)

class CheckoutViewModel(
    private val repository: Repository,
    private val cart: CartStore
) : ViewModel() {

    private val _state = MutableStateFlow(CheckoutState())
    val state: StateFlow<CheckoutState> = _state.asStateFlow()

    init {
        validasi()
    }

    fun validasi() {
        val baris = cart.isi()
        val outletId = cart.outletId()

        if (baris.isEmpty() || outletId == null) {
            _state.value = CheckoutState(memuat = false, keranjangKosong = true)
            return
        }

        _state.value = _state.value.copy(
            memuat = true,
            galat = null,
            baris = baris,
            masalah = emptyList(),
            alasan = null,
            pesanPenolakan = null
        )

        viewModelScope.launch {
            when (val hasil = repository.validasiCheckout(outletId, baris.map { it.kePayload() })) {
                is GatewayResult.Gagal -> {
                    _state.value = _state.value.copy(memuat = false, galat = hasil.error)
                }
                is GatewayResult.Sukses -> {
                    val r = hasil.data
                    if (r.ok) {
                        // Angka yang ditampilkan adalah angka gateway. Menghitung
                        // ulang di aplikasi berisiko menampilkan total yang
                        // berbeda dari yang benar-benar ditagih.
                        _state.value = _state.value.copy(
                            memuat = false,
                            subtotal = r.subtotal?.roundToLong(),
                            potongan = r.discountAmount?.roundToLong(),
                            total = r.total?.roundToLong(),
                            masalah = emptyList(),
                            alasan = null,
                            pesanPenolakan = null
                        )
                    } else {
                        // HTTP 200 dengan `ok: false` adalah PENOLAKAN, bukan
                        // keberhasilan. Total sengaja dikosongkan supaya tidak
                        // ada jalan menuju pembayaran dari keadaan ini.
                        _state.value = _state.value.copy(
                            memuat = false,
                            subtotal = null,
                            potongan = null,
                            total = null,
                            masalah = r.masalah ?: emptyList(),
                            alasan = r.alasan,
                            pesanPenolakan = pesanUntukAlasan(r.alasan, r.pesan)
                        )
                    }
                }
            }
        }
    }

    /**
     * Menjalankan pemulihan untuk satu masalah, lalu memvalidasi ulang.
     *
     * Keranjang TIDAK pernah dibuang seluruhnya: hanya item bermasalah yang
     * disentuh. Pelanggan memilih item lainnya dengan sengaja, dan membuang
     * semuanya karena satu item habis berarti menghukumnya atas keadaan
     * outlet.
     */
    fun perbaiki(masalah: CartProblemDto) {
        if (masalah.jenis == "harga_berubah" && masalah.hargaBaru != null) {
            cart.perbaruiHarga(masalah.menuItemId, masalah.hargaBaru.roundToLong())
        } else {
            cart.hapusMenuItem(masalah.menuItemId)
        }
        validasi()
    }

    /** Memperbaiki seluruh masalah sekaligus, dengan aturan yang sama. */
    fun perbaikiSemua() {
        _state.value.masalah.forEach { masalah ->
            if (masalah.jenis == "harga_berubah" && masalah.hargaBaru != null) {
                cart.perbaruiHarga(masalah.menuItemId, masalah.hargaBaru.roundToLong())
            } else {
                cart.hapusMenuItem(masalah.menuItemId)
            }
        }
        validasi()
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Checkout/CheckoutViewModel.swift
import Foundation
import Observation

/// Padanan `CheckoutViewModel.kt`. Dibuat BARU setiap kali layar checkout
/// dibuka: harga & ketersediaan bisa berubah di antara dua kunjungan, dan
/// validasi basi di titik bayar justru paling berbahaya.
@MainActor
@Observable
final class CheckoutViewModel {
    private(set) var memuat = true
    private(set) var galat: GatewayError?
    private(set) var baris: [CartLine] = []
    /// Semua nilai uang di bawah berasal dari GATEWAY, bukan hitungan sendiri.
    private(set) var subtotal: Int64?
    private(set) var potongan: Int64?
    private(set) var total: Int64?
    private(set) var masalah: [CartProblemDto] = []
    private(set) var alasan: String?
    private(set) var pesanPenolakan: String?
    private(set) var keranjangKosong = false

    /// Boleh lanjut membayar hanya kalau gateway benar-benar meloloskannya.
    var bolehLanjut: Bool { !memuat && galat == nil && total != nil && masalah.isEmpty && alasan == nil }

    @ObservationIgnored private let repository: Repository
    @ObservationIgnored private let cart: CartStore
    /// Menandai validasi terbaru: hasil validasi lama yang datang terlambat
    /// (mis. pelanggan menekan dua perbaikan beruntun) dibuang, bukan menimpa.
    @ObservationIgnored private var generasi = 0

    init(repository: Repository, cart: CartStore) {
        self.repository = repository
        self.cart = cart
    }

    func validasi() async {
        let isi = cart.isi()
        guard !isi.isEmpty, let outletId = cart.outletId() else {
            // Setara `CheckoutState(memuat=false, keranjangKosong=true)` di Android:
            // SEMUA dikosongkan, termasuk total lama — tak boleh tersisa jalan ke bayar.
            generasi += 1
            memuat = false
            keranjangKosong = true
            galat = nil
            baris = []
            subtotal = nil
            potongan = nil
            total = nil
            masalah = []
            alasan = nil
            pesanPenolakan = nil
            return
        }

        generasi += 1
        let milikku = generasi
        keranjangKosong = false
        memuat = true
        galat = nil
        baris = isi
        masalah = []
        alasan = nil
        pesanPenolakan = nil

        let hasil = await repository.validasiCheckout(outletId: outletId, items: isi.flatMap { $0.kePayloadList() })
        guard milikku == generasi else { return }

        memuat = false
        switch hasil {
        case .gagal(let g):
            galat = g
        case .sukses(let r) where r.ok:
            // Angka gateway yang ditampilkan — menghitung ulang berisiko
            // menampilkan total yang berbeda dari yang benar-benar ditagih.
            subtotal = r.subtotal.map(bulatkanRupiah)
            potongan = r.discountAmount.map(bulatkanRupiah)
            total = r.total.map(bulatkanRupiah)
        case .sukses(let r):
            // HTTP 200 + `ok: false` = PENOLAKAN. Total dikosongkan supaya
            // tidak ada jalan menuju pembayaran dari keadaan ini.
            subtotal = nil
            potongan = nil
            total = nil
            masalah = r.masalah ?? []
            alasan = r.alasan
            pesanPenolakan = pesanUntukAlasan(r.alasan, pesanDariGateway: r.pesan)
        }
    }

    /// Memulihkan SATU masalah lalu memvalidasi ulang. Keranjang tidak pernah
    /// dibuang seluruhnya — hanya item bermasalah yang disentuh.
    func perbaiki(_ m: CartProblemDto) async {
        terapkan(m)
        await validasi()
    }

    func perbaikiSemua() async {
        masalah.forEach(terapkan)
        await validasi()
    }

    private func terapkan(_ m: CartProblemDto) {
        if m.jenis == "harga_berubah", let harga = m.hargaBaru {
            cart.perbaruiHarga(menuItemId: m.menuItemId, hargaBaru: bulatkanRupiah(harga))
        } else {
            cart.hapusMenuItem(m.menuItemId)
        }
    }
}
